src/converse-httpFileUpload.js

```diff
@@ -0,0 +1,269 @@
+/*
+    The MIT License (MIT)
+
+    Copyright (c) 2014 Klaus Herberth <klaus@jsxc.org>
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+
+/**
+ * Implements Http File Upload (XEP-0363)
+ *
+ * @see {@link http://xmpp.org/extensions/xep-0363.html}
+ */
+(function (root, factory) {
+    define([
+        "converse-core",
+    ], factory);
+}(this, function (
+        converse
+    ) {
+    "use strict";
+    const { $msg, Backbone, Strophe, _, b64_sha1, moment, utils } = converse.env;
+
+    Strophe.addNamespace('HTTPUPLOAD', 'urn:xmpp:http:upload');
+
+    var requestSlotUrl; 
+    var ready;
+    var httpUploadOption = {
+        enable: true
+    }
+
+    converse.plugins.add('converse-httpFileUpload', {
+
+        dependencies: ["converse-chatboxes", "converse-disco"],
+
+        initialize() {
+            const { _converse } = this,
+                { __ } = _converse;
+            var connection = _converse.connection;
+            var domain;
+            var file;
+            var chatBox;
+            
+            _converse.FileUpload = Backbone.NativeView.extend({
+                /**
+                * Set up http file upload.
+                * 
+                * @param {*} connection the current strophe-connection
+                */
+                initFiletransfer () {
+                    connection = _converse.connection;
+                    domain = _converse.connection.domain; 
+
+                    if (httpUploadOption && requestSlotUrl != undefined) {
+                        ready = true;
+                        return;
+                    }
+                    this.discoverUploadService();
```

Comment:
I see this method gets called when the file upload button is clicked, but only in this method is the check made to see whether the server at all supports file upload.

What we instead want to do is we want to know whether file upload is supported before we at all render the file upload button in the toolbar.

I therefore think that with proper refactoring the `initFileTransfer` method can be removed entirely.